_posts/2024-09-05-digging-into-struct-initialization-performance.md

```diff
@@ -0,0 +1,251 @@
+Blake3   1.85M (539.81ns) (± 1.22%)  80.0B/op        fastest
+```
+
+And… performance is crashing down.
+
+## What is going on?
+
+We still only allocate 80B in the HEAP because structs are allocated on the stack, so what’s going on? Aren’t wrapping structs supposed to be free abstractions in Crystal? Because this is clearly not the case here. Is LLVM failing to inline the struct and method calls, and is BLAKE3 so fast that any overhead degrades performance?
+
+There are no more hints to understand what’s happening here. We must dig into the generated code to do any further investigation. I generated the LLVM IR for the above benchmark: 
+
+```
+crystal build --release --emit llvm-ir --no-debug bench.cr
+```
+
+The above command will generate a `bench.ll` file in the current directory. We build in release mode so that LLVM will optimize the code, and tell Crystal to not generate debug information to improve the readability. Sadly, I couldn't find anything weird there.
+
+> **NOTE:**
+> While I’m writing this blog post I notice that the LLVM IR does exhibit the exact same issue as the disassembly below. I have no idea how I didn’t notice it… maybe I forgot `--release` 🤦
+
+Let’s go deeper and inspect the generated assembly for my CPU. We can disassemble the executable with `objdump -d` for example, and this time I immediately noticed something weird: the function calls `Blake3::Hasher.new` followed by pages of MOV instructions, while the disassembly of the direct C function calls are just a few instructions and function calls.
+
+Here is the disassembly for the slow `struct` case. Looking into the rest of the disassembly, the same is happening inside `Blake3Hasher.new` when initializing `@hasher`: too many repeated MOV instructions to count.
+
+```
+000000000003cf70 <~procProc(Nil)@bench.cr:35>:
+   (... snip...)
+   3cfa0:       e8 0b 1d 00 00          callq  3ecb0 <*Blake3Hasher::new:Blake3Hasher>
+   3cfa5:       48 8b 84 24 10 16 00    mov    0x1610(%rsp),%rax
+   3cfac:       00
+   3cfad:       48 89 84 24 00 07 00    mov    %rax,0x700(%rsp)
+   3cfb4:       00
+   3cfb5:       48 8b 84 24 08 16 00    mov    0x1608(%rsp),%rax
+   3cfbc:       00
+   (... snip: the above 2 MOV are repeated with a different index ...)
+   3ec3b:       4c 8d bc 24 28 07 00    lea    0x728(%rsp),%r15
+   3ec42:       00
+   3ec43:       4c 89 ff                mov    %r15,%rdi
+   3ec46:       48 8b b4 24 10 07 00    mov    0x710(%rsp),%rsi
+   3ec4d:       00
+   3ec4e:       48 8b 94 24 08 07 00    mov    0x708(%rsp),%rdx
+   3ec55:       00
+   3ec56:       e8 e5 32 01 00          callq  51f40 <blake3_hasher_update>
+   (... snip ...)
+```
+
+Here is the disassembly for the optimized version (the inlined C calls). It’s so small that I don’t need to snip anything. Even if you don’t know assembly (I don’t know much myself): it reserves space on the stack (0x7b0), saves/restores callee-saved registers and calls the functions:
```

Comment:
**polish:** "optimized version" sounds weird because we haven't really optimized anything. Maybe call it the "original version"? If you would use different names for the different iterations of `blake3_hexstring`, it would be easy to reference the specific version.

Reply:
"Release build" is better :+1: 

LLVM inlines everything into the block, so you still need to search the block by line number in the disassembly.